api/envoy/api/v2/core/base.proto

```diff
@@ -157,9 +157,56 @@ message HeaderValueOption {
   // Header name/value pair that this option applies to.
   HeaderValue header = 1 [(validate.rules).message.required = true];
 
-  // Should the value be appended? If true (default), the value is appended to
-  // existing values.
-  google.protobuf.BoolValue append = 2;
+  // Action to take when evaluating duplicate header keys.
+  enum HeaderAction {
+    option (gogoproto.goproto_enum_prefix) = false;
+    // Append new value of header to existing value.
+    APPEND = 0;
+    // Replace existing header value with new value.
+    REPLACE = 1;
+    // Ignore new header value and preserve existing value.
+    PRESERVE = 2;
+  }
+  oneof duplicate_action {
+    // Should the value be appended? If true (default), the value is appended to
+    // existing values.
+    //
+    // .. attention::
+    //
+    //   **This field is deprecated**. Set the
+    //   :ref:`action<envoy_api_field_core.HeaderValueOption.action>` field instead.
+    google.protobuf.BoolValue append = 2 [deprecated = true];
+
+    // Action to take when duplicate header keys are encountered. By default,
+    // new values will be appended to the existing ones.
+    // TODO: Make this field required when not using oneof.
+    HeaderAction action = 3 [(validate.rules).enum.defined_only = true];
+  }
+}
+
+message HeaderOption {
+  // Header value that this option applies to.
+  string header = 1 [(validate.rules).string = {min_bytes: 1, max_bytes: 16384}];
+
+  // Action to take when evaluating duplicate header keys.
+  enum HeaderAction {
+    option (gogoproto.goproto_enum_prefix) = false;
+    // Remove the existing header if present.
+    REMOVE = 0;
+    // Ignore subsequent header key additions and preserve removal.
+    PRESERVE = 1;
+  }
+  // Action to take when duplicate header keys are encountered. By default,
+  // key will be removed once and subsequent additions will add the header back.
+  HeaderAction action = 2
+      [(validate.rules).message.required = true, (validate.rules).enum.defined_only = true];
+}
+
+// Custom headers to add/remove.
+message CustomHeaders {
+  repeated HeaderValueOption add = 1 [(validate.rules).repeated .max_items = 1000];
+
+  repeated HeaderOption remove = 2;
```

Comment:
Should this have a max_items validation as well?

Reply:
I don't believe so. It would appear the intent was to limit the number of appended headers since browsers have a practical limitation in the size of all the headers but this limitation isn't important when removing headers since they won't get sent in requests/responses. What do you think @zuercher ?

I'm also not 100% sure how 1000 was picked. I'm open to hearing feedback on that number as well.

Reply:
IIRC @alyssawilk put this in place so she can comment.

Reply:
IIRC (possibly incorrectly) I'd tried to limit this and gotten push back, but I'd be very happy to have limits here as well.  Make it so!  :-)

Reply:
💯